rfcs/0008-leaderless-dns.md

```diff
@@ -0,0 +1,238 @@
+
+# Motivation
+[motivation]: #motivation
+
+Currently, the DNS record controller has limited functionality for a single cluster, and none for multiple clusters without OCM. Having the DNSRecord controller enabled to manage DNS in these situations significantly eases onboarding users to the advantages of the DNSRecord controller.
+
+# Diagrams
+[diagrams]: #diagrams
+
+- [Overview](./0008-distributed-dns-assets/distributed-dns-usecases.jpg)
+- [P2P Health Checks](./0008-distributed-dns-assets/distributed-dns-healthchecks.jpg)
+- [DNS Pruning](./0008-distributed-dns-assets/distributed-dns-pruning.jpg)
+- [DNS Controller](./0008-distributed-dns-assets/distributed-dns-dns-operator.jpg)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Single Cluster
+In a single cluster scenario, create a DNS Policy that references a DNS provider. When a listener host defined in the gateway is found that matches one of the available zones in the provider, the addresses of that gateway will be created as A or CNAME records for that host in the appropriate zone file following the configuration outlined in the DNS Policy.
+
+### Single Cluster Health Checks
+If health checks are defined in the DNS Policy, these will also be created and used to remove unhealthy IPs from the Zone file, should all health checks fail, then all A records will be published to avoid an NXDOMAIN response.
+
+## Multiple Cluster
+In a multiple cluster scenario, every cluster requires a DNS Policy configured with a provider secret that specifies a unique `multiClusterID`, they should all reference the same zone(s) in the same DNS Provider for the same host. In each cluster, when a matching listener host defined in the gateway is found that matches one of the available zones in the provider the addresses from that gateway will be appended to the zone file - in a non-destructive manner - allowing multiple clusters to all add their gateways addresses to the same zone file, without destroying each other's values.
+
+### Multiple Cluster Health Checks
+If health checks are defined in the DNS Policy, these will be created on all clusters, there are then 2 types of health checks:
+1) Local Health Checks - used to check the local gateway's IP addresses are responding for the expected host.
+2) Remote Health Checks - used to check that the other IP addresses in the zone file are responding for that zones host.
+
+When a local health check is failing, this IP will not be published to the zone. If it has already been published it will be removed only if there is more than one available address
+
+When a remote health check is failing a TXT record is added/amended in the zone to note the failure - should the percentage of failing clusters exceed the `MultiClusterFailureThreshold` (configured in the DNS Policy), that unhealthy IP is removed from the zone by any cluster with an unhealthy check on that IP.
```

Comment:
could we expand on why a local check doesn't count towards this total failure number? The local check is using  the lb ip so will be going out and back in. It seems a reasonable check to count towards the failure number?

Reply:
When the local health check fails, it will immediately remove the bad IP, which means the health checks will be removed on remote clusters until the local cluster sees it is healthy and re-adds it.

Reply:
hmm ok. I recall you mentioning that before. Does that sit well with the user experience? If I set a health check and say "remove the ip if >50% of health checks fail. Then only the local health check fails and it gets removed. This could be hard to reason about?

Reply:
yes, I see this, I suppose my thinking was that this keeps local health checks the same in both scenarios, but it does sound a little difficult to explain why some health checks work one way, and some another. I'll update this so they all work as if they are remote health checks.

Reply:
update RFC now.